### PR TITLE
feat(zone): `once` is `provide` for zones

### PR DESCRIPTION
**Pull request by erights**
As suggested at https://github.com/Agoric/agoric-sdk/pull/7116/files#r1141536384

### CODE REVIEW

**Comment by erights** [comment]
With a minimal token test for the added `once` feature, I'm declaring this ready for review. (But extremely non-urgent!)

I was surprised that zones had zero tests. This test is little more than a placeholder only for `once`. But it's not clear how `once` should be tested. Reviewers, I'll drive further testing by your suggestions. Thanks.

**Comment by michaelfig** [comment]
I think both `once` has confusing semantics, just like `provide` does. I would intuitively expect that `once` can be called as many times as you want (just like `pthread_once` can be called many times), but only initializes the key if `once` has not been called before for that key in the subZone (or any of that subZone's ancestors), returning the initialized value.

What do you think?

**Comment by erights** [comment]
Happy to use a better short name. Suggestions?

**Comment by michaelfig** [comment]
> Happy to use a better short name. Suggestions?

`load(key, ifNotFoundThunk)`?

**Comment by erights** [comment]
My initial reaction is positive. Let's sleep on it.

File: packages/zone/src/once.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * `once` will simply return it. If not, then `makeValue(key)` is called to
+ * determine the initial value of that slot, which will normally be preserved
+ * by similar calls to `once` in future incarnations --- though that will be
+ * up to them.
+ *
+ * For single incarnation zones, `zone.once(key, makeValue)` just calls
+ * `makeValue(key)` and returns the result.
+ *
+ * Default zones do no uniqueness checking. It is up to the caller not to call
+ * `once` more than once per subzone,key pair within any vat incarnation.
```
Comment (mhofman) [inline packages/zone/src/once.js:18]: Why not perform uniqueness checks for non-durable zones?

File: packages/zone/src/virtual.js [hunk 2]
```diff
@@ -46,6 +47,8 @@ const defineVirtualExo = (
 export const detachedVirtualStores = Far('virtualStores', {
   detached: () => detachedVirtualStores,
   isStorable: canBeDurable,
+  once: alwaysOnce,
```
Comment (mhofman) [inline packages/zone/src/virtual.js:50]: Should a virtual once enforce that the value produced by the maker is virtual compatible?

**Review by mhofman — COMMENTED** [review]
Quick pass, just wondering about the differing semantics between heap/virtual/durable zones.

File: packages/zone/src/virtual.js (hunk 2 above)
Comment (erights) [reply packages/zone/src/virtual.js:50]: What Passables are not virtual compatible?

File: packages/zone/src/once.js (hunk 1 above)
Comment (erights) [reply packages/zone/src/once.js:18]: @michaelfig has a plan for that.

@michaelfig my inclination is to put this PR into Draft and wait for your pervasively uniqueness checking zone. Or even close this one in favor of your's. Does that seem plausible to you?

File: packages/zone/src/virtual.js (hunk 2 above)
Comment (mhofman) [reply packages/zone/src/virtual.js:50]: Where is `Passable` enforced? that's my question.

File: packages/zone/src/once.js (hunk 1 above)
Comment (michaelfig) [reply packages/zone/src/once.js:18]: > Or even close this one in favor of yours.

That's plausible.  Before that, I wanted to understand if you are also in favour of the rename I like inspired by @gibson042 -- `zone.makeOnce(key, maker)`.

Comment (erights) [reply packages/zone/src/once.js:18]: Yes. Go for it.

File: packages/zone/src/virtual.js (hunk 2 above)
Comment (erights) [reply packages/zone/src/virtual.js:50]: Got it. I am not sure it is consistently enforced. So yes, the virtual-compatible test could just be a Passable test.

**Comment by erights** [comment]
Converted to Draft for now.

To be closed in favor of @michaelfig 's alternative.

**Comment by erights** [comment]
Closing in favor of https://github.com/Agoric/agoric-sdk/pull/7891

**Pull request closed without merging**